[exporter/syslog] Add syslog exporter

kasia-kujawa opened the pull request:
**Description:** Adding syslog exporter implementation

**Link to tracking Issue:** #17982

**Testing:** unit tests, manual tests with syslog server

**Documentation:** please see README.md file

kasia-kujawa committed:
[exporter/syslog] Add syslog exporter

Signed-off-by: Katarzyna Kujawa <kkujawa@sumologic.com>

Co-authored-by: Raj Nishtala <rnishtala@sumologic.com>

kkujawa-sumo commented:
This pull request is ready for review :) 
cc: @astencel-sumo @djaglowski

astencel-sumo commented on exporter/syslogexporter/README.md:
```diff
@@ -0,0 +1,111 @@
+# Syslog Exporter
+
+**Stability level**: Alpha
```
We need a Markdown table here that lists the stability, pipeline types and distros like every component has, see example [here](https://github.com/open-telemetry/opentelemetry-collector-contrib/blob/3191d9cc96585002e2ab2d6a739c42e3f9e5fe2f/exporter/zipkinexporter/README.md#L3-L7).

We should probably start with [Development](https://github.com/open-telemetry/opentelemetry-collector/blob/main/README.md#development) stability, not Alpha.

kkujawa-sumo replied:
changed

astencel-sumo commented on exporter/syslogexporter/README.md:
```diff
@@ -0,0 +1,111 @@
+
+**Stability level**: Alpha
+
+## About The Exporter
```
Do we need this header? Maybe just drop it? I think the other components don't have the "About" section specifically called out.

kkujawa-sumo replied:
changed

astencel-sumo commented on exporter/syslogexporter/README.md:
```diff
@@ -0,0 +1,111 @@
+
+## Configuration
+
+**The following are a few configuration options available to forward syslog messages**:
```
> The following are a few configuration options

Does this mean these are not all of them? :sweat_smile:

```suggestion
**The following configuration options are available**:
```

kkujawa-sumo replied:
changed

astencel-sumo commented on exporter/syslogexporter/config.go:
```diff
@@ -0,0 +1,91 @@
+// Copyright 2023, OpenTelemetry Authors
```
```suggestion
// Copyright The OpenTelemetry Authors
```

kkujawa-sumo replied:
changed

astencel-sumo commented on exporter/syslogexporter/config.go:
```diff
@@ -0,0 +1,91 @@
+		invalidFields = append(invalidFields, errUnsupportedPort)
+	}
+
+	if !net.IsFQDN(cfg.Endpoint) && !net.IsIPAddr(cfg.Endpoint) && cfg.Endpoint != "localhost" {
```
Do we need to validate the format of the endpoint? I don't see other components doing this, and removing this would save us from adding another dependency (github.com/THREATINT/go-net with 10 stars on GitHub).

kkujawa-sumo replied:
changed

astencel-sumo commented on exporter/syslogexporter/README.md:
```diff
@@ -0,0 +1,111 @@
+  This ensures that all the syslog message headers are populated with the expected values.
+- Not using the `syslog_parser` will result in the syslog message being populated with default header values.
+
+## Configuration
```
The `Configuration` section only mentions the `retry_on_failure` or the `sending_queue` settings in the example. I believe these settings and their enabled status should be described in the list of all configuration settings.

kkujawa-sumo replied:
I added information about `retry_on_failure` and `sending_queue` to README.md

astencel-sumo commented on exporter/syslogexporter/config.go:
```diff
@@ -0,0 +1,91 @@
+// Validate the configuration for errors. This is required by component.Config.
+func (cfg *Config) Validate() error {
+	invalidFields := []error{}
+	if cfg.Port < 1 || cfg.Port > 65525 {
```
Shouldn't this be 65535?

```suggestion
	if cfg.Port < 1 || cfg.Port > 65535 {
```

kkujawa-sumo replied:
chnaged

kkujawa-sumo commented:
@atoulme @djaglowski I went through steps described in https://github.com/open-telemetry/opentelemetry-collector-contrib/blob/main/CONTRIBUTING.md#adding-new-components and I have some doubts related to two steps:

> Run make crosslink to update intra-repository dependencies. It will add a replace directive to go.mod file of every intra-repository dependant. This is necessary for your component to be included in the contrib executable. 

I didn't see any changes after executing `make crosslink`

> All components must be included in [internal/components/](https://github.com/open-telemetry/opentelemetry-collector-contrib/blob/main/internal/components) and in the respective testing harnesses. To align with the test goal of the project, components must be testable within the framework defined within the folder. If a component can not be properly tested within the existing framework, it must increase the non testable components number with a comment within the PR explaining as to why it can not be tested.

Should I add syslog exporter to `internal/components/` in this pull request or it is better/expected to do it in another pull request?

astencel-sumo commented on exporter/syslogexporter/factory.go:
```diff
@@ -0,0 +1,66 @@
+	qs := exporterhelper.NewDefaultQueueSettings()
+	qs.Enabled = false
+
+	return &Config{
```
This is not related to this pull request specifically, but I stumbled upon the fact that the TLS settings apparently do not have a `NewDefaultSettings` method similar to `NewDefaultRetrySettings` or `NewDefaultQueueSettings`. Perhaps it's worth adding such a method on the [TLSSetting](https://github.com/open-telemetry/opentelemetry-collector/blob/v0.73.0/config/configtls/configtls.go) struct?

This is not a problem now, because the default TLS config is equal to an empty, non-initialized struct. But if one day this changes, it would be useful to have the `NewDefaultSettings` method as a central place to change this.

I wonder what the @open-telemetry/collector-approvers think.

astencel-sumo commented on exporter/syslogexporter/syslog.go:
```diff
@@ -0,0 +1,197 @@
+
+const emptyValue = "-"
+
+type Syslog struct {
```
Why is this named `Syslog`? It seems a name like `sender` or `syslogSender` would be a better name, as it is concerned with sending the syslog data to a remote endpoint?

Also I don't think this struct needs to be exported (i.e. start with a capital letter)?

kkujawa-sumo commented on exporter/syslogexporter/README.md:
```diff
@@ -0,0 +1,111 @@
+
+**The following are a few configuration options available to forward syslog messages**:
+
+- `endpoint` - (default = `host.domain.com`) syslog endpoint (FQDN or IP address)
```
I changed it

astencel-sumo replied:
You probably also want to remove the `DefaultEndpoint` const from `config.go` (it's still being used in the `createDefaultConfig` function in `factory.go`).

astencel-sumo commented on exporter/syslogexporter/config.go:
```diff
@@ -0,0 +1,90 @@
+	// Syslog Port
+	DefaultPort = 514
+	// Syslog Endpoint
+	DefaultEndpoint = "host.domain.com"
```
This should probably be removed?

kkujawa-sumo replied:
removed

astencel-sumo commented on .github/dependabot.yml:
```diff
@@ -1097,8 +1102,3 @@ updates:
     schedule:
       interval: "weekly"
       day: "wednesday"
-  - package-ecosystem: "gomod"
```
Why is this being removed?

kkujawa-sumo replied:
There was problem reporter in checks:
```
diff --git a/.github/dependabot.yml b/.github/dependabot.yml
index bce0138..775dd60 100644
--- a/.github/dependabot.yml
+++ b/.github/dependabot.yml
@@ -267,6 +267,11 @@ updates:
     schedule:
       interval: "weekly"
       day: "wednesday"
+  - package-ecosystem: "gomod"
+    directory: "/exporter/syslogexporter"
+    schedule:
+      interval: "weekly"
+      day: "wednesday"
   - package-ecosystem: "gomod"
     directory: "/exporter/tanzuobservabilityexporter"
     schedule:
@@ -1097,8 +1102,3 @@ updates:
     schedule:
       interval: "weekly"
       day: "wednesday"
-  - package-ecosystem: "gomod"
-    directory: "/testbed"
-    schedule:
-      interval: "weekly"
-      day: "wednesday"
dependabot.yml is out of date, please run "make gendependabot" and commit the changes in this PR.
Error: Process completed with exit code 1.
```
https://github.com/open-telemetry/opentelemetry-collector-contrib/actions/runs/4437454266/jobs/7811020637
so I run "make gendependabot" and commited the changes in this PR.

astencel-sumo commented on exporter/syslogexporter/syslog.go:
Please rename the file too

astencel-sumo commented on exporter/syslogexporter/exporter.go:
```diff
@@ -0,0 +1,154 @@
+		params,
+		cfg,
+		s.pushLogsData,
+		exporterhelper.WithTimeout(exporterhelper.TimeoutSettings{Timeout: 0}),
```
Shouldn't we expose the timeout as a configuration option?

kasia-kujawa committed:
[exporter/syslog] make functions unexported

problem reported by golangci-lint
sender.go:57:71: unexported-return: exported func Connect returns unexported type *syslogexporter.sender, which can be annoying to use
(revive)
func Connect(logger *zap.Logger, cfg *Config, tlsConfig *tls.Config) (*sender, error) {

kasia-kujawa committed:
[exporter/syslog] make timeout configurable

kasia-kujawa committed:
[exporter/syslog] lower protocol string

kasia-kujawa committed:
[exporter/syslog] add examples with syslog receiver

kasia-kujawa committed:
[exporter/syslog] update to v0.74.0 otelcol core

astencel-sumo commented on exporter/syslogexporter/sender_test.go:
```diff
@@ -0,0 +1,72 @@
+		},
+	}
+
+	expectedForm := "\\<165\\>1 2003-08-24T05:14:15-07:00 192\\.0\\.2\\.1 myproc 8710 - " +
```
Why do you use regular expression here instead of raw strings? Is it because the order of the fields is undefined?

kkujawa-sumo replied:
yes, we cannot predict the order of elements in structured data

kasia-kujawa committed:
[exporter/syslog] change formatting of sender_test.go

andrzej-stencel reviewed (approved):
This is a good start. I would like to see some improvements, but it's probably best to include them in follow-up PRs instead of blocking this one forever.

Things that come to my mind:
- Include some examples in docs - input vs. output.
- Document the default values being used when data is missing (like default priority or default facility).
- Perhaps allow users to change the defaults with the configuration? Or maybe it's not needed, users can use processors for this.
- It seems that performing a separate network request for each Resource Log is not needed - perhaps all ResourceLogs could be sent in one request?
- Also not sure if/why we need to close the network connection after each request (with the call to `sender.close()` in the `sender.sendSyslogs()` method).

I think these are all details that can be worked on in separate PRs, if needed.

Thank you @kkujawa-sumo for the great work and for your patience with me :pray: :sweat_smile:

astencel-sumo commented:
@djaglowski can you take a look? Can you enable to CI workflows to see test results?

djaglowski reviewed (commented):
This looks good to me. Just one outstanding question.

I think we should determine which additional work should be done before moving the component to alpha, and we should create tracking issues for those things.

One thing I would like to see is an integration test between this component and the syslog receiver, but this can be in another PR. The easiest way to do it is probably by adding it to the testbed.

astencel-sumo commented on exporter/syslogexporter/config.go:
```diff
@@ -0,0 +1,89 @@
+	Endpoint string `mapstructure:"endpoint"`
+	// Syslog server port
+	Port int `mapstructure:"port"`
+	// Protocol for syslog communication
+	// options: tcp, udp
+	Protocol string `mapstructure:"protocol"`
+	// Format of syslog messages
+	Format string `mapstructure:"format"`
```
Yeah we discussed this with Kasia, the incosistency is far from ideal. Kasia was willing to change the names in the exporter to match the receiver (so use `protocol` for the RFC and something else for the network protocol). My opinion was that the exporter's naming is actually better, which is probably highly subjective. :sweat_smile: After all, a "protocol" is just an "agreed-on form of communication" in the widest sense, so I guess the word fits both uses.

Given your feedback, I'm OK with using `protocol` for the RFC and `network` for TCP/UDP. @kkujawa-sumo what do you think?

astencel-sumo replied:
I stumbled upon the [Carbon receiver](https://github.com/open-telemetry/opentelemetry-collector-contrib/blob/v0.75.0/receiver/carbonreceiver/README.md) using `transport` name for this configuration option (TCP or UDP). Whether this is a better name, I don't know :sweat_smile:

djaglowski replied:
👍 Either `network` or `transport` seems fine to me.

kkujawa-sumo replied:
Parameters are changed.

kasia-kujawa committed:
[exporter/syslog] change configuration
- replace protocol with network
- replace format with protocol

astencel-sumo commented on exporter/syslogexporter/factory.go:
```diff
@@ -0,0 +1,66 @@
+const (
+	// The value of "type" key in configuration.
+	typeStr        = "syslog"
+	stabilityLevel = component.StabilityLevelAlpha
```
Shouldn't this be Development? README says it's Development, not Alpha.

kkujawa-sumo replied:
I updated this pull request to use the new mechanism.

kasia-kujawa committed:
[exporter/syslog] change stability level to development

kasia-kujawa committed:
[exporter/syslog] update to v0.76.1

kasia-kujawa committed:
[exporter/syslog]  Update to use generated status

kkujawa-sumo commented:
@astencel-sumo @djaglowski please take a look once again ;) 
I created following issues which must be done before giving this exporter alpha status:
- https://github.com/open-telemetry/opentelemetry-collector-contrib/issues/21245
- https://github.com/open-telemetry/opentelemetry-collector-contrib/issues/21242
- https://github.com/open-telemetry/opentelemetry-collector-contrib/issues/21244

astencel-sumo commented on exporter/syslogexporter/README.md:
```diff
@@ -0,0 +1,63 @@
+[syslog_receiver]: https://github.com/open-telemetry/opentelemetry-collector-contrib/tree/main/receiver/syslogreceiver
+[filelog_receiver]: https://github.com/open-telemetry/opentelemetry-collector-contrib/tree/main/receiver/filelogreceiver
+[cryptoTLS]: https://github.com/golang/go/blob/518889b35cb07f3e71963f2ccfc0f96ee26a51ce/src/crypto/tls/common.go#L706-L709
+[development]: https://github.com/open-telemetry/opentelemetry-collector#development
```
This line is probably not needed anymore, as the link is being autogenerated together with the status table.
```suggestion
```

djaglowski replied:
@kkujawa-sumo, please take note of this. It can be done in a follow up PR.

astencel-sumo commented:
Not sure where the check errors come from:
- [check-collector-module-version](https://github.com/open-telemetry/opentelemetry-collector-contrib/actions/runs/4891495598/jobs/8785722737?pr=19647):
  Incorrect version "v0.76.1" of "go.opentelemetry.io/collector " is included in "./exporter/syslogexporter/go.mod". It should be version "v0.76.2-0.20230502195822-4df44379e094".
- [checks](https://github.com/open-telemetry/opentelemetry-collector-contrib/actions/runs/4891495598/jobs/8785722983?pr=19647):
  'status' has invalid keys: pipelines
  'status.stability' expected a map, got 'string'

Perhaps you need to merge the latest changes from `main`? 🤔

dmitryax commented:
@kkujawa-sumo please rebase, run make gotidy and update metadata.yaml to the latest schema

kasia-kujawa committed:
[exporter/syslog] update to v0.77.0 otelcol core

kkujawa-sumo commented:
@dmitryax All checks passed :)  could you merge this pull request?
